pkg/controller/deployment/deployment_controller.go

```diff
@@ -375,6 +376,47 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return nil
 	}
 	d := *obj.(*extensions.Deployment)
+
+	// If the deployment times out, we will need to add a failed condition in its
+	// status (if it's not already there), scale down all of its replication
+	// controllers, and finally update its status to reflect zero replicas.
+	if dc.timedOut(d) {
+		if !deploymentutil.HasFailed(&d) {
+			if d.Status.Conditions == nil {
+				d.Status.Conditions = []extensions.DeploymentCondition{}
+			}
+			cond := newCondition(extensions.DeploymentFailed, "TimedOut", "Deployment has timed out")
+			d.Status.Conditions = append(d.Status.Conditions, cond)
+			_, err := dc.expClient.Deployments(d.Namespace).UpdateStatus(&d)
+			return err
+		}
+
+		if d.Status.UpdatedReplicas > 0 || d.Status.Replicas > 0 {
```

Comment:
Agreed, I was ambivalent about this change anyway ... We can come back to this later.